core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPathfindingB.kt

```diff
@@ -0,0 +1,289 @@
+    val comfort: RollingStock.Comfort?,
+    val startTime: Double,
+    val steps: List<STDCMStep>,
+    val blockAvailability: BlockAvailabilityInterface,
+    val timeStep: Double,
+    val maxDepartureDelay: Double,
+    val maxRunTime: Double,
+    val tag: String?,
+    val standardAllowance: AllowanceValue?,
+    val pathfindingTimeout: Double = 120.0,
+    var graph: STDCMGraph? = null,
+) {
+
+    private var edgeToLength: EdgeToLength<STDCMEdge, STDCMEdge>? = null
+    private var estimateRemainingDistance: List<AStarHeuristic<STDCMEdge, STDCMEdge>>? = ArrayList()
+    private var starts: Set<STDCMEdge> = HashSet()
+
+    // TODO: maybe those could be local to the core function?
+    private var seen = HashMap<STDCMEdge, Int>()
+    private var queue = PriorityQueue<STDCMEdge>()
+
+    init {
+        graph =
+            STDCMGraph(
+                fullInfra,
+                rollingStock,
+                comfort,
+                timeStep,
+                blockAvailability,
+                maxRunTime,
+                startTime,
+                steps,
+                tag,
+                standardAllowance
+            )
+        edgeToLength = EdgeToLength { edge -> edge.length.cast() }
+
+        // TODO: figure out what should and should not be in the init
+    }
+
+    fun findPath(): STDCMResult? {
+        assert(steps.size >= 2) { "Not enough steps have been set to find a path" }
+
+        // Initialize the A* heuristic
+        val locations = steps.stream().map(STDCMStep::locations).toList()
+        val remainingDistanceEstimators = makeHeuristics(fullInfra, locations)
+        estimateRemainingDistance = makeAStarHeuristic(remainingDistanceEstimators, rollingStock)
+
+        val endBlocks = steps.last().locations.map { it.edge }
+        starts =
+            convertLocations(
+                graph!!,
+                steps[0].locations,
+                startTime,
+                maxDepartureDelay,
+                rollingStock,
+                endBlocks
+            )
+        val path = findPathImpl() ?: return null
+
+        return STDCMPostProcessingB(graph!!)
+            .makeResult(
+                fullInfra.rawInfra,
+                path,
+                startTime,
+                graph!!.standardAllowance,
+                rollingStock,
+                timeStep,
+                comfort,
+                maxRunTime,
+                blockAvailability
+            )
+    }
+
+    private fun findPathImpl(): Result? {
+        checkParameters()
+        for (location in starts) {
+            val len = edgeToLength!!.apply(location)
+            val totalCostUntilEdge = computeTotalCostUntilEdge(location, len)
+            val distanceLeftEstimation =
+                // TODO: inconsistent, same as Pathfinding
+                estimateRemainingDistance!![0].apply(location, Offset(0.meters))
+            location.weight = distanceLeftEstimation + totalCostUntilEdge
+            queue.add(location)
```

Comment:
The naming is a little inconsistent, generally locations are specific points on an edge.

We can generally assume that we handle the edge *start* implicitly when not stating otherwise, but that could be explained by a comment somewhere.